tsdb/exemplar.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// single circular buffer for all exemplars
+type CircularExemplarStorage struct {
+	lock      sync.RWMutex
+	index     map[string]int
+	exemplars []circularBufferEntry
```

Comment:
How about this being `[]*circularBufferEntry`? We only allocate the struct that we use, else it's just an empty pointer